docs/tutorial/typesfuns.rst

```diff
@@ -661,6 +661,93 @@ without any explicit use of ``Force`` or ``Delay``:
     ifThenElse True  t e = t;
     ifThenElse False t e = e;
 
+Codata Types
+============
+
+Codata types are like regular data types, except that they allow for us to
+define infinite data structures.  More precisely, for a type ``T``, each of its
+constructor arguments of type ``T`` are transformed into a coinductive parameter
+``Inf T``. This makes each of the ``T`` arguments lazy, and allows infinite data
+structures of type ``T`` to be built. One example of a codata type is Stream,
+which is defined as follows.
+
+.. code-block:: idris
+
+    codata Stream : Type -> Type where
+      (::) : (e : a) -> Stream a -> Stream a
+
+This gets translated into the following by the compiler.
+
+.. code-block:: idris
+
+    codata Stream : Type -> Type where
```

Comment:
Ok I made some tests. So for reasons that are not known to me, codata are
sugar for self recursive infinite data structures only.
Whether this definition is good or not, i have no experience to say.

If the word "codata" is used for the general case of infinite data
structures, then it creates confusion.

So I wouldn't create an example of codata with non-self recursion and say
that this doesn't work because it doesn't work by Definition.

On Fri, Feb 26, 2016 at 12:06 AM, Leif Warner <notifications@github.com>
wrote:

> In docs/tutorial/typesfuns.rst
> <https://github.com/idris-lang/Idris-dev/pull/2988#discussion_r54173314>:
>
> > +define infinite data structures.  More precisely, for a type ``T``, each of its
> > +constructor arguments of type ``T`` are transformed into a coinductive parameter
> > +``Inf T``. This makes each of the ``T`` arguments lazy, and allows infinite data
> > +structures of type ``T`` to be built. One example of a codata type is Stream,
> > +which is defined as follows.
> > +
> > +.. code-block:: idris
> > +
> > +    codata Stream : Type -> Type where
> > +      (::) : (e : a) -> Stream a -> Stream a
> > +
> > +This gets translated into the following by the compiler.
> > +
> > +.. code-block:: idris
> > +
> > +    codata Stream : Type -> Type where
>
> I think you can just say it translates into data Stream ...? I mean, you
> could write the impl of Stream with data + Inf instead of codata and the
> end result would be the same?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idris-lang/Idris-dev/pull/2988/files#r54173314>.
>